services/src/machine_learning/xgboost_training.rs

```diff
@@ -0,0 +1,115 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+use typetag::serde;
+
+use serde_json::Value;
+use snafu::{ensure, ResultExt};
+use xgboost_rs::{Booster, DMatrix};
+
+use crate::error::Result;
+use crate::machine_learning::xg_error as XgModuleError;
+use crate::machine_learning::MachineLearningFeature;
+use ordered_float::NotNan;
+
+use super::MachineLearningAggregator;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct XgboostTrainingParams {
+    pub no_data_value: f32, // FIXME: remove?
+    pub training_config: HashMap<String, String>,
```

Comment:
use a datatype with the concrete params?

Reply:
we shouldn't use a data type from an external crate as part of our API, because then our API would change when the crate changes. Also everything in our API should be camelCase. So can you add a new struct for this?

Reply:
fdfc62d4

commit msg:
refactor: add dedicated xgboost config builder module

Reply:
I've changed the implementation now such that the api is expected to pass a hashmap with key:value pairs used as config_parameter:value pairs. 
The config builder will then use the `From<HashMap>` impl to generate a config that should be safe to use. Later on the config builder is responsible for generating the config that is passed over the FFI.

9a5acc7a11f355e030bbdc3956967c66a323f890

Reply:
ok, wow, thanks for putting in the work. I didn't expect you to map this all out.

1. How can we be sure this covers all the parameters correctly, and that all parameters are covered?
2. I am not sure if you are using the derive_builder Builder correctly. This should auto generate the setters for the fields. Why do you implement them explicitly then?